src/services/exit_order.py

```diff
@@ -1,24 +1,334 @@
+
+from eth_typing import Address
+from web3.types import EventData
+
+from src.constants import FAR_FUTURE_EPOCH
+from src.modules.submodules.typings import ChainConfig
+from src.providers.consensus.typings import Validator
 
 from src.typings import BlockStamp
+from src.web3py.extentions.lido_validators import LidoValidator, NodeOperator, NodeOperatorIndex, StakingModuleId, \
+    NodeOperatorId
+from src.web3py.typings import Web3
+
+
+@dataclass
+class NodeOperatorPredictableState:
+    predictable_validators_total_age: int
+    predictable_validators_count: int
+    targeted_validators: int | None
+    delayed_validators: int
 
 
 class ValidatorsExit:
     """
-    Exit order is:
-    1. if NO active_keys < NO current keys
-    2. Stacking > 1% remove by sum validators age
-    3. By active validators numbers
+    Exit order predicates sequence:
+    1. Validator whose operator with the lowest number of delayed validators
+    2. Validator whose operator with the highest number of targeted validators to exit
+    3. Validator whose operator with the highest stake weight
+    4. Validator whose operator with the highest number of validators
+    5. Validator with the lowest activation epoch
+    6. Validator with the lowest index
     """
-    def __init__(self, blockstamp: BlockStamp, w3: Web3,  max_size: int):
-        self.blockstamp = blockstamp
+    staking_module_id: dict[Address, StakingModuleId]
+    lido_node_operator_stats: dict[NodeOperatorIndex, NodeOperatorPredictableState]
+    total_predictable_validators_count: int
+
+    def __init__(
+        self,
+        w3: Web3,
+        blockstamp: BlockStamp,
+        c_conf: ChainConfig,
+        exitable_lido_validators: list[LidoValidator],
+        max_validators_to_exit: int,
+    ):
         self.w3 = w3
+        self.blockstamp = blockstamp
+        self.c_conf = c_conf
+        self.validator_delayed_timeout_in_slots = Web3.to_int(
+            self.w3.lido_contracts.oracle_daemon_config.functions.get('VALIDATOR_DELAYED_TIMEOUT_IN_SLOTS').call(
+                block_identifier=blockstamp.block_hash)
+        )
+
+        self.exitable_lido_validators = exitable_lido_validators
+        self.max_validators_to_exit = max_validators_to_exit
+        self.left_queue_count = 0
+
+        # -- State init part --
+
+        operators = self.w3.lido_validators.get_lido_node_operators(blockstamp)
+        operator_validators = self.w3.lido_validators.get_lido_validators_by_node_operators(blockstamp)
 
-        # Iterator should contain only next validators
-        self.max_size = max_size
+        self.staking_module_id = {
+            operator.staking_module.staking_module_address: operator.staking_module.id
+            for operator in operators
+        }
+
+        self.lido_node_operator_stats = self._prepare_lido_node_operator_stats(
+            blockstamp, operators, operator_validators
+        )
+        self.total_predictable_validators_count = len([
```

Comment:
Missed it last time. For the Lido validators we also take into account those in flight (new deposits), it seems to be correct to take them into account here as well

Reply:
@avsetsin Do you mean that new deposits are already counted in `operator.total_deposited_validators` and we should count `pending` validators (which `activation_epoch` = `FAR_FUTURE_EPOCH`) in `self.total_predictable_validators_count` ?

Reply:
fixed

Reply:
I guess we need to calculate them somehow when we collect operators statistics

Reply:
Caught it. Fixed